### PR TITLE
simple working hybrid example of patchAPI 

### PR DESCRIPTION
**Pull request by ianamason**
Using dyninstAPI to create the PatchMgr and PatchObject.  

I should point out that I have not succeeded in getting a pacthAPI example working that
uses only the patchAPI.

### CODE REVIEW

File: patchAPI/example/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+To build you need to set the three environment variables `DYNINST`, `BOOST`, and
+`TBB` to the install directories of these three dependencies.
+
+
+For example, if you `spack` installed `dyninst` you could do:
+
+
+```
+export DYNINST=`spack location --install-dir dyninst`
```
Comment (hainest) [inline patchAPI/example/README.md]: What are these being used for? Do we need one for _all_ dependencies (e.g., elfutils, libiberty, etc.)?

File: patchAPI/example/main.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  Patcher patcher(patchMgr);
+
+  NoopSnippet::Ptr snippet = NoopSnippet::create(new NoopSnippet);
+
+  vector<PatchFunction*> functions;
+
+  binobj->funcs(back_inserter(functions));
+
+
+  for(vector<PatchFunction*>::iterator funIter = functions.begin(); funIter != functions.end(); funIter++){
```
Comment (hainest) [inline patchAPI/example/main.cpp]: C++11's range-based `for` makes this much easier to use.

File: patchAPI/example/mutatee/main.c [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+int main(int argc, const char *argv[])
+{
+  if (argc > 2){
+    foo0();
+  } else {
+    foo1();
+    foo2();
+  }
+  message("Done\n");
+  return status;
```
Comment (hainest) [inline patchAPI/example/mutatee/main.c]: Returning non-zero indicates failure to the shell. Is that relevant here?

File: patchAPI/example/main.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    PatchFunction *fun = *funIter;
+
+    vector<Point*> f_entryPoints;
```
Comment (hainest) [inline patchAPI/example/main.cpp]: Destruction of this vector leaks all of its containing objects because they are pointers.

**Comment by hainest** [comment]
@mxz297 I have a couple of comments about the code, but we can merge this without testing once those are addressed.

File: patchAPI/example/README.md (hunk 1 above)
Comment (mxz297) [reply patchAPI/example/README.md]: I think without your RPATH fixes, users will have to set these environment variables to specify where to find the installed dependencies. And depending on what are pre-installed on the system, you may need to specify paths to find BOOST, TBB, ELFUTILS, and so on.

I don't think it is worthwhile to tweak these environment variables right now. After your CMake & RPath fixes, we can clear the README file.

File: patchAPI/example/main.cpp (hunk 4 above)
Comment (mxz297) [reply patchAPI/example/main.cpp]: These objects pointed by the pointers are constructed inside Dynins  and the objects are stored in a map. These objects will be destructed when Dyninst destructs the map. Users should not explicitly destruct these objects.

Comment (ianamason) [reply patchAPI/example/main.cpp]: I did valgrind it, but got so much noise that I left it at that.
```
==1932== LEAK SUMMARY:
==1932==    definitely lost: 139,464 bytes in 449 blocks
==1932==    indirectly lost: 218,810 bytes in 2,894 blocks
==1932==      possibly lost: 110,365 bytes in 1,721 blocks
==1932==    still reachable: 1,284,088 bytes in 16,184 blocks
==1932==                       of which reachable via heuristic:
==1932==                         newarray           : 528 bytes in 1 blocks
==1932==                         multipleinheritance: 239,304 bytes in 767 blocks
==1932==         suppressed: 0 bytes in 0 blocks
==1932== Reachable blocks (those to which a pointer was found) are not shown.
==1932== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==1932==
==1932== For counts of detected and suppressed errors, rerun with: -v
==1932== ERROR SUMMARY: 62 errors from 62 contexts (suppressed: 0 from 0)

```

File: patchAPI/example/mutatee/main.c (hunk 3 above)
Comment (ianamason) [reply patchAPI/example/mutatee/main.c]: I fixed this, plus added a remark about what the rewritten program depends on
(i.e. the dyninst runtime library).

**Comment by mxz297** [comment]
@hainest Are you OK with the new changes? We can update the C++11's range-based `for` later.

**Pull request merged**